blogs/2021/09/06/bracket-pair-colorization.md

```diff
@@ -0,0 +1,400 @@
+During these 10 seconds of processing, the extension host process burns at 100% CPU and all features that are powered by extensions, such as auto-completion or diagnostics, stop functioning. Luckily, [VS Code's architecture](https://code.visualstudio.com/api/advanced-topics/extension-host#stability-and-performance)
+ensures that the UI remains responsive and documents can still be saved to disk!
+
+CoenraadS was aware of this performance issue and spent a great amount of effort on increasing speed and accuracy in version 2 of the extension by reusing the token and bracket parsing engine from VS Code. However, VS Code's API and extension architecture was not designed to allow for high performance bracket pair colorization when hundreds of thousands of bracket pairs are involved. Thus, even in Bracket Pair Colorizer 2, it takes some time until the colors reflect the new nesting levels after inserting `{` at the beginning of the file:
+
+![Extension needs more than 10 seconds to process text changes in checker.ts](./checker_ts-extension.gif)
+
+While we would have loved to just improve the performance of the extension (which certainly would have required introducing more advanced APIs, optimized for high-performance scenarios), the asynchronous communication between the renderer and the extension-host severely limits how fast bracket pair colorization can be. In particular, bracket pair colors should not be requested asynchronously as soon as they appear in the viewport, as this would have caused visible flickering. A discussion of this can be found [here](https://github.com/microsoft/vscode/issues/128465#issuecomment-879089188).
+
+### What We Did
+
+Instead, [in the 1.60 update](https://code.visualstudio.com/updates/v1_60#_high-performance-bracket-pair-colorization), we re-implemented the extension in the core of VS Code and could bring this time down to less than a millisecond - in this particular example, that is more than 10,000 times faster!
+
+The feature can be enabled by adding the setting `"editor.bracketPairColorization.enabled": true`.
+
+Now, updates are no longer noticeable, even for files with hundreds of thousands of bracket pairs! Notice how the bracket-color in line 42,788 reflects the new nesting level immediately after typing `{` in line 2:
+
+![Native implementation needs less than a millisecond to process text changes in checker.ts](./checker_ts-native.gif)
+
+Without being limited by public API design, we could use (2,3)-trees, recursion-free tree-traversal, bit-arithmetic, incremental parsing and other techniques to reduce the extension's worst-case update complexity (i.e. the time required to process user-input when a document already has been opened) from $\mathcal{O}(N + E)$ to $\mathcal{O}(\mathrm{log}^3 N + E)$ with $N$ being the document size and $E$ the edit size, assuming the nesting level of bracket pairs is bounded by $\mathcal{O}(\mathrm{log} N)$.
```

Comment:
Is there a link we can add for people who are not familiar with the big-O notation and what that means?

Reply:
I see that this is explained below too, so maybe it can be removed from here? We could just say that we implemented more advanced algorithms (and their names) which greatly reduced the complexity of the code and improved the performance. And then say "read more below for the details"

Reply:
For me, this is the abstract of the blog post and the time-complexity reduction of the update operation is the main result. That is the thing that (hopefully) motivates interested readers (who know about big O notation) to learn how we did it.

I would not say we reduced the complexity of the algorithm though :D Thanks for pointing out that I should say "time-complexity" here.
I'll link to [wikipedia](https://en.wikipedia.org/wiki/Time_complexity).